reservedinstancecheck/reservedinstancecheck.py

```diff
@@ -0,0 +1,479 @@
+
+
+class ReservationDisplayTerminal(ReservationDisplay):
+    @property
+    def reservation_table(self):
+        return "\n".join([
+            "",
+            AsciiTable(self.reservation_data,
+                       "Reservations ({})".format(self.checker.region)).table
+        ])
+
+    @property
+    def unreserved_table(self):
+        return "\n".join([
+            "",
+            AsciiTable(self.unreserved_data,
+                       "Unreserved ({})".format(self.checker.region)).table
+        ])
+
+    @property
+    def unused_table(self):
+        return "\n".join([
+            "",
+            AsciiTable(self.unused_data,
+                       "Unused ({})".format(self.checker.region)).table
+        ])
+
+
+class ReservationChecker(object):
+    @classmethod
+    def fetch_active_instance_reservations(cls, client, filters=[]):
+        """
+        Finds all active instance reservations
+
+        :param client: boto3 EC2 client
+        :type client: botocore.client.EC2
+        :param filters: Optional additional list of filters
+        :type filters: list
+        :return: list of instance reservations
+        :rtype: list[dict]
+        """
+        filters.append({
+            'Name': 'state',
+            'Values': ['active']
+        })
+        resp = client.describe_reserved_instances(Filters=filters)
+
+        for ri in resp['ReservedInstances']:
+            assert ri['Scope'] == 'Availability Zone', \
+                "Unsupported reservation scope: {}".format(ri['Scope'])
+
+        return resp['ReservedInstances']
+
+    def __init__(self, unreserved_days=None, region=None):
+        """
+        Instantiate class to recon used/unused instance reservations
+
+        :param unreserved_days: Number of days to consider an instance unreserved
+        :type unreserved_days: int
+        :param region: Explicit EC2 region to examine. Uses config as default
+        :type region: None|str
+        """
+        self.unreserved_days = unreserved_days
+        self._reservations = None
+        self._unreserved = None
+        self._unused = None
+
+        # Client setup
+        client_args = {}
+
+        if region is not None:
+            client_args.update({
+                'region_name': region,
+            })
+
+        # Instantiate client
+        self.ec2 = boto3.client('ec2', **client_args)
+
+    @property
+    def region(self):
+        return self.ec2.meta.region_name
+
+    def _find_unused_or_unreserved(self):
+        """
+        Filters through all instances to propogate `unused` and `unreserved`
+        """
+        # Determine what AZs to look through for instances
+        reserved_azs = [ri['AvailabilityZone'] for ri in self.reservations]
+        log.debug("Filtering through AZ: {}".format(reserved_azs))
+
+        # Construct a request to find all EC2 instances that are not stopped or
+        # terminated
+        pager = self.ec2.get_paginator('describe_instances')
```

Comment:
As of 99f92476b8ed7cbf601deb9001238ce5859b9962 I've updated to employ the `ServiceResource` and it's associated models rather than the straight client.

One of the reasons I didn't go that route initially is because the flat client supports the retrieval of instance reservations whereas the `ServiceResource` doesn't.